src/watch.rs

```diff
@@ -0,0 +1,123 @@
+    Sources,
+    Code,
+}
+
+pub fn watch(
+    args: ArgMatches,
+    watch_sources: HashSet<PathBuf>,
+    watch_code: HashSet<PathBuf>,
+) -> Fallible {
+    trigger_on_change(&watch_sources, &watch_code, |change| {
+        println!(
+            "{} changed. Re-building...",
+            if change == ChangeType::Sources {
+                "Sources"
+            } else {
+                "Code"
+            }
+        );
+
+        if change == ChangeType::Sources {
+            let curr_dir = env::current_dir()?;
+            let (_config, _watch_forward, _watch_reverse) =
+                cmd::run_with_args(&args, Some(change == ChangeType::Code))?;
+            env::set_current_dir(&curr_dir)?;
+        }
+
+        Ok(())
+    })?;
+
+    Ok(())
+}
+
+/// Calls the closure when a book source file is changed, blocking indefinitely.
+pub fn trigger_on_change<F>(
+    watch_sources: &HashSet<PathBuf>,
+    watch_code: &HashSet<PathBuf>,
+    closure: F,
+) -> Fallible
+where
+    F: Fn(ChangeType) -> Fallible,
+{
+    use notify::RecursiveMode::*;
+
+    let (tx_sources, rx_sources) = channel();
+    let mut source_watcher = notify::watcher(tx_sources, Duration::from_secs(1))?;
+    let (tx_code, rx_code) = channel();
+    let mut code_watcher = notify::watcher(tx_code, Duration::from_secs(1))?;
+
+    for path in watch_sources {
+        source_watcher.watch(path, NonRecursive)?;
+    }
+    for path in watch_code {
+        code_watcher.watch(path, NonRecursive)?;
+    }
+
+    let (tx_changes_sources, rx_changes) = channel();
+    let tx_changes_code = tx_changes_sources.clone();
+    std::thread::spawn(move || loop {
+        let deadline = Instant::now() + Duration::from_millis(COLLECT_EVENTS_MILLIS);
+        let mut send_event = false;
+        loop {
+            let timeout = deadline.saturating_duration_since(Instant::now());
+            if timeout.as_nanos() == 0 {
+                break;
+            }
+
+            if let Ok(event) = rx_sources.recv_timeout(timeout) {
+                if is_file_change(event) {
+                    send_event = true;
+                }
+            }
+        }
+        if send_event {
+            tx_changes_sources.send(ChangeType::Sources).unwrap();
+        }
+    });
+
+    std::thread::spawn(move || loop {
+        let deadline = Instant::now() + Duration::from_millis(COLLECT_EVENTS_MILLIS);
+        let mut send_event = false;
+        loop {
+            let timeout = deadline.saturating_duration_since(Instant::now());
+            if timeout.as_nanos() == 0 {
+                break;
+            }
+
+            if let Ok(event) = rx_code.recv_timeout(timeout) {
+                if is_file_change(event) {
+                    send_event = true;
+                }
+            }
+        }
+        if send_event {
+            tx_changes_code.send(ChangeType::Code).unwrap();
+        }
+    });
+
+    loop {
+        let event = rx_changes.recv().unwrap();
+        closure(event)?;
```

Comment:
The closure API seems overkill at this point. Why not just return `ex_changes`?

Reply:
I think I don't understand that. What is the sense of returning from the loop?

Reply:
I mean that instead of passing a closure in and calling in that in a loop, do not include the loop in the function at all. Just return the receiver and do
```rust
for event in rx_changes.iter() { ... }
```
on the outside.

Reply:
I tried to return `rx_changes` from `trigger_on_change`, and use a loop as above in `watch`. However, the cannel then seemy colsed and I get a panic at `watch.rs`, L89 (`let event = in_channel.recv().unwrap();`).

Reply:
Yes, because the watchers are kept on stack as it is. You can move those onto the threads that handle them using a binding like
```rust
thread::spawn(move || {
    let _watcher = watcher;
...
```